Powertoys Run | VSCodeWorkspaces- add support for vscode 1.64 - bug fix #15247

ricardosantos9521 opened the pull request:
## Summary of the Pull Request

**What is this about:**
VSCode insiders 1.64 has a new file for workspaces. This is pull request is to add support to the new file.

**What is included in the PR:** 
Use the new file Backups/workspaces.json instead of storage.json

**How does someone test / validate:** 
- Install lastest version of vscode-insiders.
- Open multiple project folders/workspaces or remote containers
- by typing { in Powertoys all the projects and workspaces from vscodeinsiders should appear in the result list

## Quality Checklist

- [x] **Linked issue:** #15247
- [x] **Communication:** I've discussed this with core contributors in the issue. 
- [ ] **Tests:** Added/updated and all pass
- [ ] **Installer:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Docs:** Added/ updated
- [ ] **Binaries:** Any new files are added to WXS / YML
   - [x] No new binaries
   - [ ] [YML for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.

ricardosantos9521 committed:
#15247 Powertoys Run | VSCodeWorkspaces- add support for vscode 1.64 new workspace.json file

stefansjfw commented on src/modules/launcher/Plugins/Community.PowerToys.Run.Plugin.VSCodeWorkspaces/WorkspacesHelper/VSCodeWorkspacesApi.cs:
```diff
@@ -70,7 +73,7 @@ public List<VSCodeWorkspace> Workspaces
                             VSCodeStorageFile vscodeStorageFile = JsonSerializer.Deserialize<VSCodeStorageFile>(fileContent);
 
-                            if (vscodeStorageFile != null)
+                            if (vscodeStorageFile != null && vscodeStorageFile.OpenedPathsList != null)
```
When VS Code v1.64 is out and installed locally (updated current version) this file will still exist & contain `OpenedPathsList`? In that case new workspaces stored in new file won't be picked up?

Also, on VS Code update, will workspaces from old storage.json file be copied to new file or not ?

ricardosantos9521 replied:
the storage.json file will exist but the OpenedPathsList will not. 
When vscode updates the workspaces from the storage.json will be moved to Backup/workspaces.json by vscode as far as i know.

ricardosantos9521 commented:
I found a issue with this solution please don't merge it.

Aaron-Junker commented:
@jaimecbernardo So a merge blocking bug was found here. So I guess we will release this not with 0.53 right?

jaimecbernardo commented:
Yes, guess this will be targeted at 0.55.

ricardosantos9521 committed:
Get previous open workspaces for sqlite file

ricardosantos9521 commented:
I think i fixed the issue. I'm obtaining the previous opened workspaces from AppData\Roaming\Code - Insiders\User\globalStorage\state.vscdb a sqlite db. I added the package Microsoft.Data.Sqlite do i need to add it somewhere else?

ricardosantos9521 committed:
check-spelling-bot Report fix

stefansjfw commented:
> I think i fixed the issue. I'm obtaining the previous opened workspaces from AppData\Roaming\Code - Insiders\User\globalStorage\state.vscdb a sqlite db. I added the package Microsoft.Data.Sqlite do i need to add it somewhere else?
> 
> @stefansjfw

yes, installer should probably be updated. Let me check

stefansjfw commented:
yes, check this line https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs#L1333

Before modifying anything, try building the PowerToys installer like this (in developer CMD for Visual Studio) and install PT:

```
nuget.exe restore
MSBuild -m .\PowerToys.sln /p:Configuration=Release /p:Platform="x64"
nuget.exe restore .\tools\BugReportTool\
MSBuild -m .\tools\BugReportTool\BugReportTool.sln /p:Configuration=Release /p:Platform="x64"
nuget.exe restore .\tools\WebcamReportTool\
MSBuild -m .\tools\WebcamReportTool\WebcamReportTool.sln /p:Configuration=Release /p:Platform="x64"
nuget.exe restore .\installer\
MSBuild -m .\installer\PowerToysSetup.sln /target:PowerToysInstaller /p:Configuration=Release /p:Platform="x64"
MSBuild -m .\installer\PowerToysSetup.sln /p:Configuration=Release /p:Platform="x64"
``` 

My guess is that when you try to test VSCodeWorkspaces plugin - part with reading the database, it will fail.

ricardosantos9521 committed:
add dlls to installer and set SqliteConnection to readonly

ricardosantos9521 commented:
@stefansjfw the installer is working now

jaimecbernardo commented:
Hi @ricardosantos9521,
I feel like the merge didn't work as expected. The changeset on github is showing changes on different files. Perhaps a merge from the 'main' branch will fix it.

ricardosantos9521 commented:
> nuget.exe restore
> MSBuild -m .\PowerToys.sln /p:Configuration=Release /p:Platform="x64"

ricardosantos9521 commented:
> Hi @ricardosantos9521, I feel like the merge didn't work as expected. The changeset on github is showing changes on different files. Perhaps a merge from the 'main' branch will fix it.

done

stefansjfw commented on src/modules/launcher/Plugins/Community.PowerToys.Run.Plugin.VSCodeWorkspaces/WorkspacesHelper/VSCodeWorkspacesApi.cs:
```diff
@@ -106,6 +110,50 @@ public List<VSCodeWorkspace> Workspaces
                                     }
                                 }
                             }
+                            else if (File.Exists(vscode_storage_db))
```
Should this be inside `if (File.Exists(vscode_storage))`. Does db logic depend on existence of `vscode_storage` file? I.e. can we still read workspaces from DB if there is no `vscode_storage` file?

ricardosantos9521 replied:
No, we can't read workspaces if there's no vscode_storage.

stefansjfw replied:
👍

stefansjfw reviewed (approved):
Other than that one question I wrote, looks good! Nice work!
